### PR TITLE
Fix NYI messages from LuaJIT

### LINKED ISSUE #27: NYI messages from LuaJIT
Hi John,

Running the `soak.lua` or the units tests with `-jv` outputs a few NYI messages. NYI messages indicate that this has not yet been implemented in the in LuaJIT's JIT compiler. See: http://wiki.luajit.org/NYI.

I think it will be good to solve these problems (where possible), especially in performance-critical code (such as `voperation` and `gvalue` and `vobject`).

NYI messages:
1. ~~`[NYI: bytecode 50 at voperation.lua:102]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/f898808373d567e6dbd501cf941f777442d50acb.
2. ~~`[NYI: bytecode 51 at voperation.lua:118]`~~ Fixed with: https://github.com/jcupitt/lua-vips/pull/28
3. ~~`[NYI: bytecode 51 at voperation.lua:188]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/8c46fac10444e987a00d401011b9b0213b11ec2d.
4. ~~`[NYI: bytecode 72 at voperation.lua:214]`~~ Not fixable, we must iterate with `pairs` here.
5. ~~`[NYI: bytecode 50 at voperation.lua:265]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/c4d87b599583ab88c87d76278036c8a4e6efabf1.
6. ~~`[NYI: unsupported C type conversion at vobject.lua:89]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/f898808373d567e6dbd501cf941f777442d50acb
7. ~~`[NYI: bytecode 51 at Image_methods.lua:771]`~~  Not fixable, we can't circumvent the vargargs closure.
8. ~~`[NYI: unsupported C type conversion at Image_methods.lua:448]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/f898808373d567e6dbd501cf941f777442d50acb.
9. `[NYI: unsupported C type conversion at gvalue.lua:134]`, see footnote 1.
10. ~~`[NYI: bytecode 51 at voperation.lua:91]`~~ Fixed with: https://github.com/kleisauke/lua-vips/commit/6dd0e10d7ad2179aea437bbe2230c8f5c958fd0c

Footnotes:
1. I think there might be ints in the array (because I don't see these messages for the `array_int_type` gtype).

I'll make a PR in the course of next week.

### PR DESCRIPTION
**Pull request by kleisauke**
Hi John,

I had some time to cook up a PR to fix the NYI messages from LuaJIT. For easier reviewing, I've split up the changes in multiple commits.

This PR closes #27.

### CODE REVIEW

**Comment by jcupitt** [comment]
Ooop!! I completely missed this, sorry @kleisauke. Duh.

I'll look it over now.

File: src/vips/Image_methods.lua [hunk 1]
```diff
@@ -185,18 +185,20 @@ function Image.new_from_array(array, scale, offset)
...
 
     local n = width * height
-    local a = ffi.new(gvalue.pdouble_typeof, n)
+    local a = {}
     for y = 0, height - 1 do
         for x = 0, width - 1 do
             a[x + y * width] = array[y + 1][x + 1]
         end
     end
+    a = ffi.new(gvalue.double_arr_typeof, n, a)
```
Comment (jcupitt) [inline src/vips/Image_methods.lua]: Oooof, changing the type of a is a bit ugly. I'll adjust this.

File: src/vips/voperation.lua [hunk 2]
```diff
@@ -95,9 +84,9 @@ local voperation_mt = {
                 if gtype == gvalue.image_type then
                     value = match_image:imageize(value)
                 elseif gtype == gvalue.array_image_type then
-                    value = map(function(x)
-                        return match_image:imageize(x)
-                    end, value)
+                    for i = 1, #value do
```
Comment (jcupitt) [inline src/vips/voperation.lua]: Getting rid of the higher-order functions is a bit of a shame, but I can see the reasoning.

File: src/vips/cdefs.lua [hunk 3]
```diff
@@ -37,51 +37,51 @@ ffi.cdef [[
 
     int vips_enum_from_nick (const char* domain,
         GType gtype, const char* str);
-    const char *vips_enum_nick (GType gtype, int value);
+    const char* vips_enum_nick (GType gtype, int value);
```
Comment (jcupitt) [inline src/vips/cdefs.lua]: I usually prefer C style, but I agree, perhaps C++ style is better.

File: src/vips/Image_methods.lua (hunk 1 above)
Comment (kleisauke) [reply src/vips/Image_methods.lua]: Hopefully fixed with https://github.com/libvips/lua-vips/pull/29/commits/d2c062f7a5d360f68810b3f894760a14ee83f46f.

File: src/vips/cdefs.lua (hunk 3 above)
Comment (kleisauke) [reply src/vips/cdefs.lua]: Yeah, I think I was trying to unify the file at that time. I switched to C-style decls with commit https://github.com/libvips/lua-vips/pull/29/commits/c342d9497d1789667df29a737ec30a038df5e3d3.

**Comment by kleisauke** [comment]
Don't worry, this dropped from my radar as well (it's been a while since I did something with Lua).

fwiw, I've also updated the CI and dependencies with commit https://github.com/libvips/lua-vips/pull/29/commits/68df35fe30c14ae6ebecc1ef678925a7e6cea441.

**Comment by jcupitt** [comment]
Argh so late. Sorry again.

**Pull request merged**